neqo-transport/src/connection.rs

```diff
@@ -2773,47 +2809,65 @@ impl LossRecovery {
 
-        if loss_time != 0 {
+        if loss_time.is_some() {
             self.loss_detection_timer = loss_time;
         } else if has_crypto_out {
-            self.loss_detection_timer = self.time_of_last_sent_crypto_packet
-                + self
+            self.loss_detection_timer = self.time_of_last_sent_crypto_packet.map(|i| {
+                i + self
                     .rtt_vals
-                    .timer_for_crypto_retransmission(self.crypto_count);
+                    .timer_for_crypto_retransmission(self.crypto_count)
+            });
         } else {
             // Calculate PTO duration
-            let timeout = self.rtt_vals.pto() * 2u64.pow(self.pto_count);
-            self.loss_detection_timer = self.time_of_last_sent_ack_eliciting_packet + timeout;
+            let timeout = self.rtt_vals.pto() * 2u32.pow(self.pto_count);
+            self.loss_detection_timer = self
+                .time_of_last_sent_ack_eliciting_packet
+                .map(|i| i + timeout);
         }
-        qdebug!([self] "loss_detection_timer={}", self.loss_detection_timer);
+        qdebug!([self] "loss_detection_timer={:?}", self.loss_detection_timer);
     }
 
-    #[allow(clippy::if_same_then_else)]
-    fn get_earliest_loss_time(&self) -> (u64, PNSpace) {
+    fn get_earliest_loss_time(&self) -> (Option<Instant>, PNSpace) {
         let mut loss_time = self.packet_spaces[PNSpace::Initial as usize].loss_time;
         let mut pn_space = PNSpace::Initial;
         for space in &[PNSpace::Handshake, PNSpace::ApplicationData] {
-            let packet_space = &self.packet_spaces[*space as usize];
-
-            if loss_time == 0 {
-                loss_time = packet_space.loss_time;
-                pn_space = *space;
-            } else if packet_space.loss_time != 0 && packet_space.loss_time < loss_time {
-                loss_time = packet_space.loss_time;
-                pn_space = *space;
+            let packet_space = self.space(*space);
+
+            if let Some(new_loss_time) = packet_space.loss_time {
+                if loss_time.is_none() || Some(new_loss_time) < loss_time {
+                    loss_time = Some(new_loss_time);
+                    pn_space = *space;
+                }
             }
         }
+
         (loss_time, pn_space)
     }
 
-    pub fn get_timer(&self) -> u64 {
-        self.loss_detection_timer * 1000
+    /// This is when we'd like to be called back, so we can see if losses have
+    /// occurred.
+    pub fn get_timer(&self, cur_time: Instant) -> Option<Duration> {
+        match self.loss_detection_timer {
+            None => None,
+            Some(time) => {
+                // if cur_time is greater than timer, then we want to be called back
+                // immediately?
+                if cur_time >= time {
+                    // TODO(agrover@mozilla.com): Look into this more
+                    // Specify smallest unit of time
+                    Some(Duration::from_nanos(0))
```

Comment:
Yeah, this needs investigation.  If there is something we should be doing *right now*, then we should have already done it.

Reply:
The recovery spec (A.8) does say something about  "This algorithm may result in the timer being set in the past, particularly if timers wake up late. Timers set in the past SHOULD fire immediately."

Should we just be ensuring the caller re-calls us, or should we go ahead and assert this never happens?